src/Microsoft.AspNetCore.Server.Kestrel/BadHttpRequestException.cs

```diff
@@ -17,6 +23,14 @@ private BadHttpRequestException(string message, int statusCode)
 
         internal int StatusCode { get; }
 
+        internal void SetAdditionalHeaders(FrameResponseHeaders headers)
```

Comment:
I would prefer subclass of `BadHttpRequestException` with `RequiredMethod` property and downcast check in `ProduceEnd`